// @shelby-protocol/sdk/dist/core/operations/generated/types.mjs
import "../../../chunk-7P6ASYW6.mjs";

// src/core/operations/generated/types.ts
var Blob_Activities_Select_Column = /* @__PURE__ */ ((Blob_Activities_Select_Column2) => {
  Blob_Activities_Select_Column2["BlobName"] = "blob_name";
  Blob_Activities_Select_Column2["EventIndex"] = "event_index";
  Blob_Activities_Select_Column2["EventType"] = "event_type";
  Blob_Activities_Select_Column2["Timestamp"] = "timestamp";
  Blob_Activities_Select_Column2["TransactionHash"] = "transaction_hash";
  Blob_Activities_Select_Column2["TransactionVersion"] = "transaction_version";
  return Blob_Activities_Select_Column2;
})(Blob_Activities_Select_Column || {});
var Blobs_Select_Column = /* @__PURE__ */ ((Blobs_Select_Column2) => {
  Blobs_Select_Column2["BlobCommitment"] = "blob_commitment";
  Blobs_Select_Column2["BlobName"] = "blob_name";
  Blobs_Select_Column2["CreatedAt"] = "created_at";
  Blobs_Select_Column2["ExpiresAt"] = "expires_at";
  Blobs_Select_Column2["IsDeleted"] = "is_deleted";
  Blobs_Select_Column2["IsWritten"] = "is_written";
  Blobs_Select_Column2["NumChunksets"] = "num_chunksets";
  Blobs_Select_Column2["Owner"] = "owner";
  Blobs_Select_Column2["PlacementGroup"] = "placement_group";
  Blobs_Select_Column2["Size"] = "size";
  Blobs_Select_Column2["SliceAddress"] = "slice_address";
  Blobs_Select_Column2["UpdatedAt"] = "updated_at";
  return Blobs_Select_Column2;
})(Blobs_Select_Column || {});
var Cursor_Ordering = /* @__PURE__ */ ((Cursor_Ordering2) => {
  Cursor_Ordering2["Asc"] = "ASC";
  Cursor_Ordering2["Desc"] = "DESC";
  return Cursor_Ordering2;
})(Cursor_Ordering || {});
var Order_By = /* @__PURE__ */ ((Order_By2) => {
  Order_By2["Asc"] = "asc";
  Order_By2["AscNullsFirst"] = "asc_nulls_first";
  Order_By2["AscNullsLast"] = "asc_nulls_last";
  Order_By2["Desc"] = "desc";
  Order_By2["DescNullsFirst"] = "desc_nulls_first";
  Order_By2["DescNullsLast"] = "desc_nulls_last";
  return Order_By2;
})(Order_By || {});
var Placement_Group_Slots_Select_Column = /* @__PURE__ */ ((Placement_Group_Slots_Select_Column2) => {
  Placement_Group_Slots_Select_Column2["PlacementGroup"] = "placement_group";
  Placement_Group_Slots_Select_Column2["SlotIndex"] = "slot_index";
  Placement_Group_Slots_Select_Column2["Status"] = "status";
  Placement_Group_Slots_Select_Column2["StorageProvider"] = "storage_provider";
  Placement_Group_Slots_Select_Column2["UpdatedAt"] = "updated_at";
  return Placement_Group_Slots_Select_Column2;
})(Placement_Group_Slots_Select_Column || {});
var Processor_Status_Select_Column = /* @__PURE__ */ ((Processor_Status_Select_Column2) => {
  Processor_Status_Select_Column2["LastSuccessVersion"] = "last_success_version";
  Processor_Status_Select_Column2["LastTransactionTimestamp"] = "last_transaction_timestamp";
  Processor_Status_Select_Column2["LastUpdated"] = "last_updated";
  return Processor_Status_Select_Column2;
})(Processor_Status_Select_Column || {});
export {
  Blob_Activities_Select_Column,
  Blobs_Select_Column,
  Cursor_Ordering,
  Order_By,
  Placement_Group_Slots_Select_Column,
  Processor_Status_Select_Column
};
