// @shelby-protocol/clay-codes/dist/index-node.js
import "./chunk-E7TRDNBU.js";
import {
  FULL_WASM_MASK,
  MAX_WASM_MASK_BITS,
  convertToErasedMask,
  makeDecoderAPI
} from "./chunk-RXWC2ROY.js";
import {
  makeEncoderAPI
} from "./chunk-BWTMZB33.js";
import "./chunk-ST33RHQN.js";
import {
  flattenSystematic
} from "./chunk-22OVODFS.js";
import {
  CLAY_PARAMS_BYTES,
  DEFAULT_ALIGNMENT,
  DEFAULT_ALIGN_MASK,
  DEFAULT_PAGE_SIZE
} from "./chunk-N26WXNR7.js";
import {
  DuplicateChunkIndexError,
  InvalidChunkIndexError
} from "./chunk-XAN62WH2.js";

// src/index-node.ts
import { readFile } from "fs/promises";
import { dirname, resolve } from "path";
import { fileURLToPath } from "url";
async function loadWasm() {
  const here = dirname(fileURLToPath(import.meta.url));
  const paths = [
    resolve(here, "clay.wasm"),
    resolve(here, "../dist/clay.wasm")
  ];
  let bytes;
  for (const path of paths) {
    try {
      bytes = await readFile(path);
      break;
    } catch {
    }
  }
  if (!bytes) {
    throw new Error(`Unable to locate clay.wasm. Tried: ${paths.join(", ")}`);
  }
  const defaultImports = {
    env: {
      abort: () => {
        throw new Error("WASM abort called");
      },
      js_msg: (i) => console.log(`wasm msg: ${i}`)
    }
  };
  const module = await WebAssembly.compile(bytes);
  return WebAssembly.instantiate(module, defaultImports);
}
async function createEncoder(opts) {
  const instance = await loadWasm();
  return makeEncoderAPI(instance, opts);
}
async function createDecoder(opts) {
  const instance = await loadWasm();
  return makeDecoderAPI(instance, opts);
}
export {
  CLAY_PARAMS_BYTES,
  DEFAULT_ALIGNMENT,
  DEFAULT_ALIGN_MASK,
  DEFAULT_PAGE_SIZE,
  DuplicateChunkIndexError,
  FULL_WASM_MASK,
  InvalidChunkIndexError,
  MAX_WASM_MASK_BITS,
  convertToErasedMask,
  createDecoder,
  createEncoder,
  flattenSystematic,
  makeDecoderAPI,
  makeEncoderAPI
};
