// @shelby-protocol/sdk/dist/core/erasure/index.mjs
import "../../chunk-XNEIWM4O.mjs";
import {
  ReedSolomonErasureCodingProvider
} from "../../chunk-NHWWORCH.mjs";
import {
  createDefaultErasureCodingProvider,
  defaultErasureCodingConfig,
  erasureCodingConfig16Total10Data13Helper,
  erasureCodingConfig4Total2Data3Helper
} from "../../chunk-XRLHLAOW.mjs";
import {
  ClayErasureCodingProvider
} from "../../chunk-OGKZ575S.mjs";
import "../../chunk-JTXYKO3U.mjs";
import {
  DEFAULT_ERASURE_D,
  DEFAULT_ERASURE_K,
  DEFAULT_ERASURE_M,
  DEFAULT_ERASURE_N,
  DEFAULT_SAMPLE_SIZE,
  ERASURE_CODE_PARAMS,
  ErasureCodingScheme
} from "../../chunk-AUQDI5BS.mjs";
import "../../chunk-7P6ASYW6.mjs";
export {
  ClayErasureCodingProvider,
  DEFAULT_ERASURE_D,
  DEFAULT_ERASURE_K,
  DEFAULT_ERASURE_M,
  DEFAULT_ERASURE_N,
  DEFAULT_SAMPLE_SIZE,
  ERASURE_CODE_PARAMS,
  ErasureCodingScheme,
  ReedSolomonErasureCodingProvider,
  createDefaultErasureCodingProvider,
  defaultErasureCodingConfig,
  erasureCodingConfig16Total10Data13Helper,
  erasureCodingConfig4Total2Data3Helper
};
