// @shelby-protocol/sdk/dist/chunk-WJTVYLKW.mjs
import {
  defaultErasureCodingConfig
} from "./chunk-XRLHLAOW.mjs";
import {
  getShelbyIndexerClient
} from "./chunk-KYQRWJ3U.mjs";
import {
  getAptosConfig
} from "./chunk-IKRWBDG5.mjs";
import {
  createBlobKey
} from "./chunk-3NRBHSMQ.mjs";
import {
  expectedTotalChunksets
} from "./chunk-BXEVML7N.mjs";
import {
  getBlobNameSuffix,
  normalizeAddress
} from "./chunk-UEZNZBJO.mjs";
import {
  ERASURE_CODE_AND_CHUNK_MAPPING
} from "./chunk-JTXYKO3U.mjs";
import {
  ERASURE_CODE_PARAMS
} from "./chunk-AUQDI5BS.mjs";
import {
  SHELBY_DEPLOYER
} from "./chunk-7OV5ZYW6.mjs";

// src/core/clients/ShelbyBlobClient.ts
import {
  AccountAddress,
  Aptos,
  Hex,
  U32
} from "@aptos-labs/ts-sdk";
var MissingTransactionSubmitterError = class extends Error {
  constructor() {
    super(
      "USD sponsorship (usdSponsor) requires a transaction submitter to be configured. The sponsored transaction is a multi-agent transaction that must be co-signed by the sponsor. Configure a transaction submitter via aptos.pluginSettings.TRANSACTION_SUBMITTER that routes transactions to a service (like a gas station) capable of providing the sponsor's signature."
    );
    this.name = "MissingTransactionSubmitterError";
  }
};
var ShelbyBlobClient = class _ShelbyBlobClient {
  aptos;
  deployer;
  indexer;
  defaultOptions;
  /**
   * The ShelbyBlobClient is used to interact with the Shelby contract on the Aptos blockchain. This
   * includes functions for registering blob commitments and retrieving blob metadata.
   *
   * @param config - The client configuration object.
   * @param config.network - The Shelby network to use.
   * @param defaultOptions - Optional default options for blob operations.
   *
   * @example
   * ```typescript
   * const blobClient = new ShelbyBlobClient({
   *   aptos: {
   *     network: Network.SHELBYNET,
   *     clientConfig: {
   *       API_KEY: "AG-***",
   *     },
   *   },
   * });
   * ```
   *
   * @example
   * ```typescript
   * // With default options for USD sponsorship
   * const blobClient = new ShelbyBlobClient(
   *   {
   *     network: Network.SHELBYNET,
   *     aptos: {
   *       pluginSettings: {
   *         TRANSACTION_SUBMITTER: myGasStationSubmitter,
   *       },
   *     },
   *   },
   *   {
   *     usdSponsor: { feePayerAddress: sponsorAddress },
   *   }
   * );
   * ```
   */
  constructor(config, defaultOptions) {
    this.aptos = new Aptos(getAptosConfig(config));
    this.deployer = config.deployer ?? AccountAddress.fromString(SHELBY_DEPLOYER);
    this.indexer = getShelbyIndexerClient(config);
    this.defaultOptions = defaultOptions ?? {};
  }
  /**
   * Merges method-level options with default options, giving precedence to method-level values.
   */
  mergeOptions(options) {
    return {
      build: options?.build ?? this.defaultOptions.build,
      submit: options?.submit ?? this.defaultOptions.submit,
      usdSponsor: options?.usdSponsor ?? this.defaultOptions.usdSponsor,
      chunksetSizeBytes: options?.chunksetSizeBytes ?? this.defaultOptions.chunksetSizeBytes
    };
  }
  /**
   * Validates that if USD sponsorship is requested, a transaction submitter is
   * configured as well. Checks both the client-level and method-level transaction
   * submitter configurations. Throws MissingTransactionSubmitterError if usdSponsor is
   * provided without a transaction submitter.
   */
  validateUsdSponsorConfig(usdSponsor, submit) {
    if (!usdSponsor) {
      return;
    }
    const hasClientSubmitter = this.aptos.config.getTransactionSubmitter();
    const hasMethodSubmitter = submit?.transactionSubmitter !== void 0 && submit?.transactionSubmitter !== null;
    if (!hasClientSubmitter && !hasMethodSubmitter) {
      throw new MissingTransactionSubmitterError();
    }
  }
  /**
   * Retrieves the blob metadata from the blockchain. If it does not exist,
   * returns `undefined`.
   *
   * @param params.account - The account namespace the blob is stored in (e.g. "0x1")
   * @param params.name - The name of the blob (e.g. "foo/bar")
   * @returns The blob metadata.
   *
   * @example
   * ```typescript
   * const metadata = await client.getBlobMetadata({
   *   account: AccountAddress.fromString("0x1"),
   *   name: "foo/bar.txt",
   * });
   * ```
   */
  async getBlobMetadata(params) {
    try {
      const rawMetadata = await this.aptos.view({
        payload: {
          function: `${this.deployer.toString()}::blob_metadata::get_blob_metadata`,
          functionArguments: [
            createBlobKey({
              account: params.account,
              blobName: params.name
            })
          ]
        }
      });
      if (!rawMetadata?.[0]?.vec?.[0]) {
        return void 0;
      }
      const metadata = rawMetadata[0].vec[0];
      let encoding;
      if (metadata.encoding.__variant__ === "ClayCode_16Total_10Data_13Helper") {
        encoding = {
          variant: "clay",
          ...ERASURE_CODE_PARAMS[metadata.encoding.__variant__],
          ...ERASURE_CODE_AND_CHUNK_MAPPING[metadata.encoding.__variant__]
        };
      } else if (metadata.encoding.__variant__ === "ClayCode_4Total_2Data_3Helper") {
        encoding = {
          variant: "clay",
          ...ERASURE_CODE_PARAMS[metadata.encoding.__variant__],
          ...ERASURE_CODE_AND_CHUNK_MAPPING[metadata.encoding.__variant__]
        };
      } else {
        throw new Error(
          "Could not parse encoding from Shelby Smart Contract, this SDK is out of date."
        );
      }
      return {
        blobMerkleRoot: Hex.fromHexInput(
          metadata.blob_commitment
        ).toUint8Array(),
        owner: normalizeAddress(metadata.owner),
        name: params.name,
        blobNameSuffix: getBlobNameSuffix(params.name),
        size: Number(metadata.blob_size),
        encoding,
        expirationMicros: Number(metadata.expiration_micros),
        creationMicros: Number(metadata.creation_micros),
        sliceAddress: normalizeAddress(metadata.slice.inner),
        isWritten: metadata.is_written
      };
    } catch (error) {
      if (error instanceof Error && // Depending on the network, the error message may show up differently.
      (error.message?.includes("sub_status: Some(404)") || error.message?.includes("EBLOB_NOT_FOUND"))) {
        return void 0;
      }
      throw error;
    }
  }
  /**
   * Retrieves all the blobs and their metadata for an account from the
   * blockchain.
   *
   * @param params.account - The account namespace the blobs are stored in (e.g. "0x1")
   * @param params.pagination (optional) - The pagination options.
   * @param params.orderBy (optional) - The order by clause to sort the blobs by.
   * @returns The blob metadata for all the blobs for the account.
   *
   * @example
   * ```typescript
   * // BlobMetadata[]
   * const blobs = await client.getAccountBlobs({
   *   account: AccountAddress.fromString("0x1"),
   * });
   * ```
   */
  getAccountBlobs(params) {
    const { where, ...rest } = params;
    return this.getBlobs({
      where: {
        ...where,
        owner: { _eq: normalizeAddress(params.account).toString() }
      },
      pagination: rest.pagination,
      orderBy: rest.orderBy
    });
  }
  /**
   * Retrieves blobs and their metadata from the blockchain.
   *
   * @param params.where (optional) - The where clause to filter the blobs by.
   * @param params.pagination (optional) - The pagination options.
   * @param params.orderBy (optional) - The order by clause to sort the blobs by.
   * @returns The blob metadata for all the blobs that match the where clause.
   *
   * @example
   * ```typescript
   * // BlobMetadata[]
   * const blobs = await client.getBlobs({
   *   where: { owner: { _eq: AccountAddress.fromString("0x1").toString() } },
   * });
   * ```
   */
  async getBlobs(params = {}) {
    const { limit, offset } = params.pagination ?? {};
    const { orderBy, where } = params;
    const currentMicros = String(Date.now() * 1e3);
    const defaultActiveFilter = {
      expires_at: { _gte: currentMicros },
      is_deleted: { _eq: "0" }
    };
    const finalWhere = where !== void 0 ? { ...defaultActiveFilter, ...where } : defaultActiveFilter;
    const { blobs } = await this.indexer.getBlobs({
      where: finalWhere,
      limit,
      offset,
      orderBy
    });
    return blobs.map(
      (blob) => ({
        owner: normalizeAddress(blob.owner),
        name: blob.blob_name,
        blobNameSuffix: getBlobNameSuffix(blob.blob_name),
        blobMerkleRoot: Hex.fromHexInput(blob.blob_commitment).toUint8Array(),
        size: Number(blob.size),
        // TODO: Add encoding when supported in NCI
        encoding: {
          variant: "clay",
          ...ERASURE_CODE_PARAMS.ClayCode_16Total_10Data_13Helper,
          ...ERASURE_CODE_AND_CHUNK_MAPPING.ClayCode_16Total_10Data_13Helper
        },
        expirationMicros: Number(blob.expires_at),
        creationMicros: Number(blob.created_at),
        sliceAddress: normalizeAddress(blob.slice_address),
        isWritten: Boolean(Number(blob.is_written)),
        isDeleted: Boolean(Number(blob.is_deleted))
      })
    );
  }
  async getBlobActivities(params) {
    const { limit, offset } = params.pagination ?? {};
    const { orderBy, where } = params;
    const { blob_activities } = await this.indexer.getBlobActivities({
      where,
      limit,
      offset,
      orderBy
    });
    const activityTypeMapping = {
      [`${this.deployer.toStringLong()}::blob_metadata::BlobRegisteredEvent`]: "register_blob",
      [`${this.deployer.toStringLong()}::blob_metadata::BlobDeletedEvent`]: "delete_blob",
      [`${this.deployer.toStringLong()}::blob_metadata::BlobExpirationExtendedEvent`]: "extend_blob_expiration",
      [`${this.deployer.toStringLong()}::blob_metadata::BlobWrittenEvent`]: "write_blob"
    };
    return blob_activities.map(
      (activity) => ({
        blobName: activity.blob_name,
        accountAddress: normalizeAddress(activity.blob_name.substring(1, 65)),
        type: activityTypeMapping[activity.event_type] ?? "unknown",
        eventType: activity.event_type,
        eventIndex: Number(activity.event_index),
        transactionHash: activity.transaction_hash,
        // Using Number here in JS is technically not okay because txn version can be
        // as large as u64::MAX, but it should be fine for a long while.
        transactionVersion: Number(activity.transaction_version),
        timestamp: `${activity.timestamp}Z`
      })
    );
  }
  /**
   * Retrieves the total number of blobs from the blockchain.
   *
   * @param params.where (optional) - The where clause to filter the blobs by.
   * @returns The total number of blobs.
   *
   * @example
   * ```typescript
   * const count = await client.getBlobsCount();
   * ```
   */
  async getBlobsCount(params) {
    const { where } = params;
    const { blobs_aggregate } = await this.indexer.getBlobsCount({ where });
    return blobs_aggregate?.aggregate?.count ?? 0;
  }
  /**
   * Retrieves the total size of blobs from the blockchain.
   *
   * @param params.where (optional) - The where clause to filter the blobs by.
   * @returns The total size of blobs in bytes.
   *
   * @example
   * ```typescript
   * const size = await client.getTotalBlobsSize();
   * ```
   */
  async getTotalBlobsSize(params = {}) {
    const { where } = params;
    const { blobs_aggregate } = await this.indexer.getTotalBlobsSize({ where });
    return Number(blobs_aggregate?.aggregate?.sum?.size ?? 0);
  }
  /**
   * Retrieves the total number of blob activities from the blockchain.
   *
   * @param params.where (optional) - The where clause to filter the blob activities by.
   * @returns The total number of blob activities.
   *
   * @example
   * ```typescript
   * const count = await client.getBlobActivitiesCount();
   * ```
   */
  async getBlobActivitiesCount(params) {
    const { where } = params;
    const { blob_activities_aggregate } = await this.indexer.getBlobActivitiesCount({ where });
    return blob_activities_aggregate?.aggregate?.count ?? 0;
  }
  /**
   * Registers a blob on the blockchain by writing its merkle root and metadata.
   *
   * @param params.account - The account that is signing and paying for the transaction.
   * @param params.blobName - The name/path of the blob (e.g. "foo/bar.txt").
   * @param params.blobMerkleRoot - The merkle root of the blob commitments.
   * @param params.size - The size of the blob in bytes.
   * @param params.expirationMicros - The expiration time of the blob in microseconds.
   * @param params.options - Optional transaction building options.
   * @param params.options.chunksetSizeBytes - Custom chunkset size (defaults to DEFAULT_CHUNKSET_SIZE_BYTES).
   * @param params.options.build - Additional Aptos transaction building options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   * const provider = await ClayErasureCodingProvider.create();
   * const blobCommitments = await generateCommitments(provider, data);
   *
   * const { transaction } = await client.registerBlob({
   *   account: signer,
   *   blobName: "foo/bar.txt",
   *   blobMerkleRoot: blobCommitments.blob_merkle_root,
   *   size: data.length,
   *   expirationMicros: Date.now() * 1000 + 3600_000_000, // 1 hour from now in microseconds
   * });
   * ```
   */
  async registerBlob(params) {
    const options = this.mergeOptions(params.options);
    this.validateUsdSponsorConfig(options.usdSponsor, options.submit);
    const config = params.config ?? defaultErasureCodingConfig();
    const chunksetSize = config.chunkSizeBytes * config.erasure_k;
    const buildArgs = {
      ...options.build,
      data: _ShelbyBlobClient.createRegisterBlobPayload({
        deployer: this.deployer,
        account: params.account.accountAddress,
        blobName: params.blobName,
        blobSize: params.size,
        blobMerkleRoot: params.blobMerkleRoot,
        numChunksets: expectedTotalChunksets(params.size, chunksetSize),
        expirationMicros: params.expirationMicros,
        useSponsoredUsdVariant: options.usdSponsor !== void 0,
        encoding: config.enumIndex
      }),
      sender: params.account.accountAddress
    };
    const transaction = options.usdSponsor ? await this.aptos.transaction.build.multiAgent({
      ...buildArgs,
      secondarySignerAddresses: [options.usdSponsor.feePayerAddress]
    }) : await this.aptos.transaction.build.simple(buildArgs);
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.account,
        transaction,
        ...options.submit
      })
    };
  }
  /**
   * Deletes a blob on the blockchain.
   *
   * @param params.account - The account that is signing and paying for the transaction.
   * @param params.blobName - The name/path of the blob (e.g. "foo/bar.txt").
   * @param params.options - Optional transaction building options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   *
   * const { transaction } = await client.deleteBlob({
   *   account: signer,
   *   blobName: "foo/bar.txt",
   * });
   * ```
   */
  async deleteBlob(params) {
    const transaction = await this.aptos.transaction.build.simple({
      ...params.options,
      data: _ShelbyBlobClient.createDeleteBlobPayload({
        deployer: this.deployer,
        blobName: params.blobName
      }),
      sender: params.account.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.account,
        transaction
      })
    };
  }
  /**
   * Deletes multiple blobs on the blockchain in a single atomic transaction.
   *
   * **Note:** This function requires the `delete_multiple_blobs` entry function
   * which will be deployed to the smart contract on 2026-02-04. Using this
   * function before that date will result in a transaction failure.
   *
   * This operation is atomic: if any blob deletion fails (e.g., blob not found),
   * the entire transaction fails and no blobs are deleted.
   *
   * @param params.account - The account that is signing and paying for the transaction.
   * @param params.blobNames - Array of blob name suffixes without the account address prefix
   *   (e.g. ["foo/bar.txt", "baz.txt"], NOT ["0x1/foo/bar.txt"]). The account address
   *   prefix is automatically derived from the signer.
   * @param params.options - Optional transaction building options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   *
   * const { transaction } = await client.deleteMultipleBlobs({
   *   account: signer,
   *   blobNames: ["foo/bar.txt", "baz.txt"],
   * });
   * ```
   */
  async deleteMultipleBlobs(params) {
    const transaction = await this.aptos.transaction.build.simple({
      ...params.options,
      data: _ShelbyBlobClient.createDeleteMultipleBlobsPayload({
        deployer: this.deployer,
        blobNames: params.blobNames
      }),
      sender: params.account.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.account,
        transaction
      })
    };
  }
  /**
   * Acks the blob chunksets on-chain. If each chunkset meets the necessary threshold, the entire blob will be marked as written.
   *
   * @param params.account - The account that is signing the transaction.
   * @param params.blobOwner - The account that owns the blob.
   * @param params.blobName - The name of the blob (e.g. "foo/bar")
   * @param params.creationMicros - The creation time of the blob in microseconds.
   * @param params.chunksetIdx - The index of the chunkset being acknowledged.
   * @param params.storageProviderChunksetAcks - The signatures
   * @param params.options - Additional options for transaction building and encoding.
   *
   * @returns The blob commitments and the pending transaction.
   *
   * @example
   * ```typescript
   * const { transaction } = await client.addChunksetAcknowledgements({
   *   account: signer,
   *   blobOwner: owner,
   *   blobName: "foo/bar.txt",
   *   creationMicros, // Taken from the blob metadata at registration time.
   *   chunksetIdx,
   *   storageProviderAcks: An array of StorageProviderAck types, each having the slot index and signature from the SP.
   * });
   * ```
   */
  async addChunksetAcknowledgements(params) {
    const transaction = await this.aptos.transaction.build.simple({
      ...params.options?.build,
      data: _ShelbyBlobClient.createChunksetAcknowledgementsPayload({
        blobOwner: params.blobOwner,
        blobName: params.blobName,
        creationMicros: params.creationMicros,
        chunksetIdx: params.chunksetIdx,
        storageProviderAcks: params.storageProviderAcks
      }),
      sender: params.account.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.account,
        transaction
      })
    };
  }
  /**
   * Registers multiple blobs on the blockchain by writing their merkle roots and metadata.
   *
   * @param params.account - The account that is signing and paying for the transaction.
   * @param params.expirationMicros - The expiration time of the blobs in microseconds.
   * @param params.blobs - The blobs to register.
   * @param params.blobs.blobName - The name/path of the blob (e.g. "foo/bar.txt").
   * @param params.blobs.blobSize - The size of the blob in bytes.
   * @param params.blobs.blobMerkleRoot - The merkle root of the blob commitments as a hex string.
   * @param params.options - Optional transaction building options.
   * @param params.options.chunksetSizeBytes - Custom chunkset size (defaults to DEFAULT_CHUNKSET_SIZE_BYTES).
   * @param params.options.build - Additional Aptos transaction building options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   * const provider = await ClayErasureCodingProvider.create();
   * const blobCommitments = await generateCommitments(provider, data);
   *
   * const { transaction } = await client.batchRegisterBlobs({
   *   account: signer,
   *   expirationMicros: Date.now() * 1000 + 3600_000_000, // 1 hour from now in microseconds
   *   blobs: [
   *     {
   *       blobName: "foo/bar.txt",
   *       blobSize: data.length,
   *       blobMerkleRoot: blobCommitments.blob_merkle_root,
   *     },
   *   ],
   * });
   * ```
   */
  async batchRegisterBlobs(params) {
    const options = this.mergeOptions(params.options);
    this.validateUsdSponsorConfig(options.usdSponsor, options.submit);
    const config = params.config ?? defaultErasureCodingConfig();
    const chunksetSize = config.chunkSizeBytes * config.erasure_k;
    const buildArgs = {
      ...options.build,
      sender: params.account.accountAddress,
      data: _ShelbyBlobClient.createBatchRegisterBlobsPayload({
        deployer: this.deployer,
        account: params.account.accountAddress,
        expirationMicros: params.expirationMicros,
        blobs: params.blobs.map((blob) => ({
          blobName: blob.blobName,
          blobSize: blob.blobSize,
          blobMerkleRoot: blob.blobMerkleRoot,
          numChunksets: expectedTotalChunksets(blob.blobSize, chunksetSize)
        })),
        useSponsoredUsdVariant: options.usdSponsor !== void 0,
        encoding: config.enumIndex
      })
    };
    const transaction = options.usdSponsor ? await this.aptos.transaction.build.multiAgent({
      ...buildArgs,
      secondarySignerAddresses: [options.usdSponsor.feePayerAddress]
    }) : await this.aptos.transaction.build.simple(buildArgs);
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.account,
        transaction,
        ...options.submit
      })
    };
  }
  /**
   * Creates a transaction payload to register a blob on the blockchain.
   * This is a static helper method for constructing the Move function call payload.
   *
   * @param params.deployer - Optional deployer account address. Defaults to SHELBY_DEPLOYER.
   * @param params.account - The account that will own the blob.
   * @param params.blobName - The name/path of the blob (e.g. "foo/bar.txt").
   * @param params.blobSize - The size of the blob in bytes.
   * @param params.blobMerkleRoot - The merkle root of the blob commitments as a hex string.
   * @param params.expirationMicros - The expiration time of the blob in microseconds.
   * @param params.numChunksets - The total number of chunksets in the blob.
   *
   * @returns An Aptos transaction payload data object for the register_blob Move function.
   *
   * @see https://github.com/shelby/shelby/blob/e08e84742cf2b80ad8bb7227deb3013398076d53/move/shelby_contract/sources/global_metadata.move#L357
   */
  static createRegisterBlobPayload(params) {
    const functionName = params.useSponsoredUsdVariant ? "register_blob_with_sponsor" : "register_blob";
    return {
      function: `${(params.deployer ?? SHELBY_DEPLOYER).toString()}::blob_metadata::${functionName}`,
      functionArguments: [
        params.blobName,
        params.expirationMicros,
        Hex.fromHexString(params.blobMerkleRoot).toUint8Array(),
        params.numChunksets,
        params.blobSize,
        // TODO
        0,
        // payment tier
        params.encoding
      ]
    };
  }
  /**
   * Creates a transaction payload to register multiple blobs on the blockchain.
   * This is a static helper method for constructing the Move function call payload.
   *
   * @param params.deployer - Optional deployer account address. Defaults to SHELBY_DEPLOYER.
   * @param params.account - The account that will own the blobs.
   * @param params.expirationMicros - The expiration time of the blobs in microseconds.
   * @param params.blobs - The blobs to register.
   * @param params.blobs.blobName - The name/path of the blob (e.g. "foo/bar.txt").
   * @param params.blobs.blobSize - The size of the blob in bytes.
   * @param params.blobs.blobMerkleRoot - The merkle root of the blob commitments as a hex string.
   * @param params.blobs.numChunksets - The total number of chunksets in the blob.
   *
   * @returns An Aptos transaction payload data object for the register_multiple_blobs Move function.
   *
   * @see https://github.com/shelby/shelby/blob/e08e84742cf2b80ad8bb7227deb3013398076d53/move/shelby_contract/sources/global_metadata.move#L357
   */
  static createBatchRegisterBlobsPayload(params) {
    const functionName = params.useSponsoredUsdVariant ? "register_multiple_blobs_with_sponsor" : "register_multiple_blobs";
    const blobNames = [];
    const blobMerkleRoots = [];
    const blobNumChunksets = [];
    const blobSizes = [];
    params.blobs.forEach((blob) => {
      blobNames.push(blob.blobName);
      blobMerkleRoots.push(
        Hex.fromHexString(blob.blobMerkleRoot).toUint8Array()
      );
      blobNumChunksets.push(blob.numChunksets);
      blobSizes.push(blob.blobSize);
    });
    return {
      function: `${(params.deployer ?? SHELBY_DEPLOYER).toString()}::blob_metadata::${functionName}`,
      functionArguments: [
        blobNames,
        params.expirationMicros,
        blobMerkleRoots,
        blobNumChunksets,
        blobSizes,
        // TODO
        0,
        // payment tier
        params.encoding
      ]
    };
  }
  /**
   * Creates a transaction payload to delete a blob on the blockchain.
   * This is a static helper method for constructing the Move function call payload.
   *
   * @param params.deployer - Optional deployer account address. Defaults to SHELBY_DEPLOYER.
   * @param params.blobName - The blob name (e.g. "bar.txt", without the account address prefix).
   *
   * @returns An Aptos transaction payload data object for the delete_blob Move function.
   *
   * @see https://github.com/shelby/shelby/blob/64e9d7b4f0005e586faeb1e4085c79159234b6b6/move/shelby_contract/sources/global_metadata.move#L616
   */
  static createDeleteBlobPayload(params) {
    return {
      function: `${(params.deployer ?? SHELBY_DEPLOYER).toString()}::blob_metadata::delete_blob`,
      functionArguments: [params.blobName]
    };
  }
  /**
   * Creates a transaction payload to delete multiple blobs on the blockchain.
   * This is a static helper method for constructing the Move function call payload.
   *
   * **Note:** This function requires the `delete_multiple_blobs` entry function
   * which will be deployed to the smart contract on 2026-02-04. Using this
   * function before that date will result in a transaction failure.
   *
   * This operation is atomic: if any blob deletion fails (e.g., blob not found),
   * the entire transaction fails and no blobs are deleted.
   *
   * @param params.deployer - Optional deployer account address. Defaults to SHELBY_DEPLOYER.
   * @param params.blobNames - Array of blob name suffixes without the account address prefix
   *   (e.g. ["foo/bar.txt", "baz.txt"], NOT ["0x1/foo/bar.txt"]). The account address
   *   prefix is automatically derived from the transaction sender.
   *
   * @returns An Aptos transaction payload data object for the delete_multiple_blobs Move function.
   *
   * @see https://github.com/shelby/shelby/blob/main/move/shelby_contract/sources/blob_metadata.move
   */
  static createDeleteMultipleBlobsPayload(params) {
    return {
      function: `${(params.deployer ?? SHELBY_DEPLOYER).toString()}::blob_metadata::delete_multiple_blobs`,
      functionArguments: [params.blobNames]
    };
  }
  static createChunksetAcknowledgementsPayload(params) {
    const ackBitMask = params.storageProviderAcks.reduce(
      (acc, ack) => acc | 1 << ack.slot,
      0
    );
    return {
      function: `${SHELBY_DEPLOYER}::blob_metadata::add_chunkset_acknowledgements`,
      functionArguments: [
        createBlobKey({
          account: params.blobOwner,
          blobName: params.blobName
        }),
        params.chunksetIdx,
        params.creationMicros,
        new U32(Number(ackBitMask)),
        params.storageProviderAcks.map((ack) => ack.signature)
      ]
    };
  }
  static createBlobAcknowledgementsPayload(params) {
    const ackBitMask = params.storageProviderAcks.reduce(
      (acc, ack) => acc | 1 << ack.slot,
      0
    );
    return {
      function: `${SHELBY_DEPLOYER}::blob_metadata::add_blob_acknowledgements`,
      functionArguments: [
        createBlobKey({
          account: params.blobOwner,
          blobName: params.blobName
        }),
        params.creationMicros,
        new U32(Number(ackBitMask)),
        params.storageProviderAcks.map((ack) => ack.signature)
      ]
    };
  }
  async addBlobAcknowledgements(params) {
    const transaction = await this.aptos.transaction.build.simple({
      ...params.options?.build,
      data: _ShelbyBlobClient.createBlobAcknowledgementsPayload({
        blobOwner: params.blobOwner,
        blobName: params.blobName,
        creationMicros: params.creationMicros,
        storageProviderAcks: params.storageProviderAcks
      }),
      sender: params.account.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.account,
        transaction
      })
    };
  }
};

export {
  MissingTransactionSubmitterError,
  ShelbyBlobClient
};
