// @shelby-protocol/sdk/dist/chunk-TVLV7C74.mjs
import {
  MICROPAYMENTS_DEPLOYER
} from "./chunk-7OV5ZYW6.mjs";

// src/core/types/payments.ts
import {
  AccountAddress,
  Deserializer,
  Hex,
  Serializer
} from "@aptos-labs/ts-sdk";
var MICROPAYMENTS_MODULE_NAME = "micropayments";
var WITHDRAW_APPROVAL_STRUCT_NAME = "WithdrawApproval";
function serializeTypeInfo(serializer, moduleAddress, moduleName, structName) {
  moduleAddress.serialize(serializer);
  const moduleNameBytes = new TextEncoder().encode(moduleName);
  serializer.serializeBytes(moduleNameBytes);
  const structNameBytes = new TextEncoder().encode(structName);
  serializer.serializeBytes(structNameBytes);
}
var StaleChannelStateError = class _StaleChannelStateError extends Error {
  /**
   * The last valid micropayment stored by the server.
   * Clients can use this to reset their local channel state.
   */
  storedMicropayment;
  constructor(storedMicropayment, message) {
    super(
      message ?? "Client has stale channel state. Use the returned micropayment to reset local state."
    );
    this.name = "StaleChannelStateError";
    this.storedMicropayment = storedMicropayment;
  }
  /**
   * Returns the stored micropayment as a base64-encoded string.
   */
  toBase64() {
    const bytes = this.storedMicropayment.bcsToBytes();
    const binaryString = Array.from(
      bytes,
      (byte) => String.fromCharCode(byte)
    ).join("");
    return btoa(binaryString);
  }
  /**
   * Creates a StaleChannelStateError from a base64-encoded micropayment string.
   */
  static fromBase64(base64, message) {
    const binaryString = atob(base64);
    const bytes = Uint8Array.from(binaryString, (char) => char.charCodeAt(0));
    const micropayment = SenderBuiltMicropayment.deserialize(bytes);
    return new _StaleChannelStateError(micropayment, message);
  }
};
var SenderBuiltMicropayment = class _SenderBuiltMicropayment {
  /**
   * The sender's address (owner of the payment channel).
   */
  sender;
  /**
   * The receiver's address (beneficiary of the withdrawal).
   */
  receiver;
  /**
   * The unique ID of the payment channel.
   */
  paymentChannelId;
  /**
   * The cumulative amount the receiver is authorized to withdraw.
   */
  amount;
  /**
   * The fungible asset metadata address.
   */
  fungibleAssetAddress;
  /**
   * Monotonically increasing sequence number for replay protection.
   */
  sequenceNumber;
  /**
   * The sender's Ed25519 public key (32 bytes).
   * Used by the receiver to verify the signature before submitting.
   */
  publicKey;
  /**
   * The Ed25519 signature of the SignedMessage<WithdrawApproval> struct.
   * The SignedMessage includes TypeInfo for domain separation.
   */
  signature;
  /**
   * The deployer address of the micropayments module.
   * This is needed to reconstruct the TypeInfo for signature verification.
   */
  deployer;
  constructor(sender, receiver, paymentChannelId, amount, fungibleAssetAddress, sequenceNumber, publicKey, signature, deployer) {
    this.sender = sender;
    this.receiver = receiver;
    this.paymentChannelId = paymentChannelId;
    this.amount = amount;
    this.fungibleAssetAddress = fungibleAssetAddress;
    this.sequenceNumber = sequenceNumber;
    this.publicKey = publicKey;
    this.signature = signature;
    this.deployer = deployer ?? AccountAddress.fromString(MICROPAYMENTS_DEPLOYER);
  }
  /**
   * Creates the BCS-serialized message that was signed.
   * This is a SignedMessage<WithdrawApproval> which includes:
   * 1. TypeInfo (module_address, module_name, struct_name)
   * 2. WithdrawApproval struct fields
   *
   * This format is used with signature_verify_strict_t for domain separation.
   */
  getSignedMessage() {
    return _SenderBuiltMicropayment.buildSignedMessage({
      deployer: this.deployer,
      sender: this.sender,
      receiver: this.receiver,
      fungibleAssetAddress: this.fungibleAssetAddress,
      amount: this.amount,
      paymentChannelId: this.paymentChannelId,
      sequenceNumber: this.sequenceNumber
    });
  }
  /**
   * Static helper to build the SignedMessage<WithdrawApproval> bytes from raw parameters.
   * This can be used to create the message for signing without constructing the full object.
   *
   * @param params - The withdrawal approval parameters.
   * @returns The BCS-serialized SignedMessage<WithdrawApproval> bytes.
   */
  static buildSignedMessage(params) {
    const serializer = new Serializer();
    serializeTypeInfo(
      serializer,
      params.deployer,
      MICROPAYMENTS_MODULE_NAME,
      WITHDRAW_APPROVAL_STRUCT_NAME
    );
    params.sender.serialize(serializer);
    params.receiver.serialize(serializer);
    params.fungibleAssetAddress.serialize(serializer);
    serializer.serializeU64(params.amount);
    serializer.serializeU64(params.paymentChannelId);
    serializer.serializeU64(params.sequenceNumber);
    return serializer.toUint8Array();
  }
  serialize(serializer) {
    this.sender.serialize(serializer);
    this.receiver.serialize(serializer);
    this.fungibleAssetAddress.serialize(serializer);
    serializer.serializeU64(this.amount);
    serializer.serializeU64(this.paymentChannelId);
    serializer.serializeU64(this.sequenceNumber);
    serializer.serializeBytes(this.publicKey);
    serializer.serializeBytes(this.signature);
    this.deployer.serialize(serializer);
  }
  bcsToBytes() {
    const serializer = new Serializer();
    this.serialize(serializer);
    return serializer.toUint8Array();
  }
  bcsToHex() {
    return Hex.fromHexInput(this.bcsToBytes());
  }
  toStringWithoutPrefix() {
    return this.bcsToHex().toStringWithoutPrefix();
  }
  toString() {
    return this.bcsToHex().toString();
  }
  /**
   * Deserializes a SenderBuiltMicropayment from BCS bytes.
   * @param bytes - The bytes to deserialize from (Uint8Array or hex string).
   * @returns A new SenderBuiltMicropayment instance.
   */
  static deserialize(bytes) {
    const bytesArray = typeof bytes === "string" ? Hex.fromHexInput(bytes).toUint8Array() : bytes;
    const deserializer = new Deserializer(bytesArray);
    const sender = AccountAddress.deserialize(deserializer);
    const receiver = AccountAddress.deserialize(deserializer);
    const fungibleAssetAddress = AccountAddress.deserialize(deserializer);
    const amount = deserializer.deserializeU64();
    const paymentChannelId = deserializer.deserializeU64();
    const sequenceNumber = deserializer.deserializeU64();
    const publicKey = deserializer.deserializeBytes();
    const signature = deserializer.deserializeBytes();
    const deployer = AccountAddress.deserialize(deserializer);
    return new _SenderBuiltMicropayment(
      sender,
      receiver,
      paymentChannelId,
      amount,
      fungibleAssetAddress,
      sequenceNumber,
      publicKey,
      signature,
      deployer
    );
  }
};

export {
  StaleChannelStateError,
  SenderBuiltMicropayment
};
