// @shelby-protocol/sdk/dist/chunk-JTXYKO3U.mjs
// src/core/chunk.ts
var ChunkSizeScheme = /* @__PURE__ */ ((ChunkSizeScheme2) => {
  ChunkSizeScheme2["ChunkSet10MiB_Chunk1MiB"] = "ChunkSet10MiB_Chunk1MiB";
  ChunkSizeScheme2["ChunkSet2MiB_Chunk1MiB"] = "ChunkSet2MiB_Chunk1MiB";
  return ChunkSizeScheme2;
})(ChunkSizeScheme || {});
var CHUNK_SIZE_PARAMS = {
  ["ChunkSet10MiB_Chunk1MiB" /* ChunkSet10MiB_Chunk1MiB */]: {
    // 1MiB
    chunkSizeBytes: 1 * 1024 * 1024,
    // 10MiB
    chunksetSizeBytes: 10 * 1024 * 1024
  },
  ["ChunkSet2MiB_Chunk1MiB" /* ChunkSet2MiB_Chunk1MiB */]: {
    // 1MiB
    chunkSizeBytes: 1 * 1024 * 1024,
    // 2MiB
    chunksetSizeBytes: 2 * 1024 * 1024
  }
};
var DEFAULT_CHUNK_SIZE_BYTES = CHUNK_SIZE_PARAMS["ChunkSet10MiB_Chunk1MiB" /* ChunkSet10MiB_Chunk1MiB */].chunkSizeBytes;
var DEFAULT_CHUNKSET_SIZE_BYTES = CHUNK_SIZE_PARAMS["ChunkSet10MiB_Chunk1MiB" /* ChunkSet10MiB_Chunk1MiB */].chunksetSizeBytes;
var ERASURE_CODE_AND_CHUNK_MAPPING = {
  ["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */]: {
    ...CHUNK_SIZE_PARAMS.ChunkSet10MiB_Chunk1MiB
  },
  ["ClayCode_4Total_2Data_3Helper" /* ClayCode_4Total_2Data_3Helper */]: {
    ...CHUNK_SIZE_PARAMS.ChunkSet2MiB_Chunk1MiB
  }
};

export {
  ChunkSizeScheme,
  CHUNK_SIZE_PARAMS,
  DEFAULT_CHUNK_SIZE_BYTES,
  DEFAULT_CHUNKSET_SIZE_BYTES,
  ERASURE_CODE_AND_CHUNK_MAPPING
};
