// @shelby-protocol/sdk/dist/chunk-IKRWBDG5.mjs
import {
  NetworkToShelbyRPCBaseUrl
} from "./chunk-7OV5ZYW6.mjs";

// src/core/clients/utils.ts
import { AptosConfig } from "@aptos-labs/ts-sdk";
var getAptosConfig = (config) => {
  const baseFaucetConfig = config.faucet?.authToken ? { AUTH_TOKEN: config.faucet.authToken } : void 0;
  if (config.aptos) {
    return new AptosConfig({
      // Spread user's aptos config first
      ...config.aptos,
      // Network from top-level is authoritative - cannot be overridden
      network: config.network,
      // Merge clientConfig with API_KEY default
      clientConfig: {
        API_KEY: config.apiKey,
        ...config.aptos.clientConfig
      },
      // Merge faucetConfig: aptos config first, then base authToken (authoritative)
      faucetConfig: {
        ...config.aptos.faucetConfig,
        ...baseFaucetConfig
      }
    });
  }
  return new AptosConfig({
    network: config.network,
    clientConfig: {
      API_KEY: config.apiKey
    },
    faucetConfig: baseFaucetConfig
  });
};
var getShelbyRPCBaseUrl = (config) => {
  return config.rpc?.baseUrl ?? NetworkToShelbyRPCBaseUrl[config.network] ?? NetworkToShelbyRPCBaseUrl.testnet;
};

export {
  getAptosConfig,
  getShelbyRPCBaseUrl
};
