// @shelby-protocol/sdk/dist/chunk-AUQDI5BS.mjs
// src/core/erasure/constants.ts
var ErasureCodingScheme = /* @__PURE__ */ ((ErasureCodingScheme2) => {
  ErasureCodingScheme2["ClayCode_16Total_10Data_13Helper"] = "ClayCode_16Total_10Data_13Helper";
  ErasureCodingScheme2["ClayCode_4Total_2Data_3Helper"] = "ClayCode_4Total_2Data_3Helper";
  return ErasureCodingScheme2;
})(ErasureCodingScheme || {});
var ERASURE_CODE_PARAMS = {
  ["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */]: {
    // total chunks (data + parity)
    erasure_n: 16,
    // data chunks
    erasure_k: 10,
    // helper nodes
    erasure_d: 13,
    // enum index
    enumIndex: 0
  },
  ["ClayCode_4Total_2Data_3Helper" /* ClayCode_4Total_2Data_3Helper */]: {
    // total chunks (data + parity)
    erasure_n: 4,
    // data chunks
    erasure_k: 2,
    // helper nodes
    erasure_d: 3,
    // enum index
    enumIndex: 1
  }
};
var DEFAULT_ERASURE_N = ERASURE_CODE_PARAMS["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */].erasure_n;
var DEFAULT_ERASURE_K = ERASURE_CODE_PARAMS["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */].erasure_k;
var DEFAULT_ERASURE_D = ERASURE_CODE_PARAMS["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */].erasure_d;
var DEFAULT_ERASURE_M = DEFAULT_ERASURE_N - DEFAULT_ERASURE_K;
var DEFAULT_SAMPLE_SIZE = 1024;

export {
  ErasureCodingScheme,
  ERASURE_CODE_PARAMS,
  DEFAULT_ERASURE_N,
  DEFAULT_ERASURE_K,
  DEFAULT_ERASURE_D,
  DEFAULT_ERASURE_M,
  DEFAULT_SAMPLE_SIZE
};
