// @shelby-protocol/sdk/dist/chunk-CQ6QPIZK.mjs
// src/core/errors.ts
var ShelbyErrorCodes = {
  // blob_metadata.move errors
  E_BLOB_NOT_FOUND: "E_BLOB_NOT_FOUND",
  E_INVALID_EXPIRATION_TIME: "E_INVALID_EXPIRATION_TIME",
  E_TEST_ENCODING_NOT_ALLOWED: "E_TEST_ENCODING_NOT_ALLOWED",
  E_BLOB_NOT_EXPIRED: "E_BLOB_NOT_EXPIRED",
  E_INVALID_CHUNKSET_COUNT: "E_INVALID_CHUNKSET_COUNT",
  E_NOT_BLOB_OWNER: "E_NOT_BLOB_OWNER",
  E_NOT_ADMIN: "E_NOT_ADMIN",
  E_BLOB_EXPIRED: "E_BLOB_EXPIRED",
  E_BLOB_NAME_TOO_LONG: "E_BLOB_NAME_TOO_LONG",
  E_INVALID_PAYMENT_EPOCHS: "E_INVALID_PAYMENT_EPOCHS",
  E_NO_SLICES_AVAILABLE: "E_NO_SLICES_AVAILABLE",
  // Common Aptos/Move errors
  EALREADY_EXISTS: "EALREADY_EXISTS"
};
function isBlobAlreadyExistsError(errorMessage) {
  return errorMessage.includes(ShelbyErrorCodes.EALREADY_EXISTS) || errorMessage.includes("already exists");
}
function isAccessDeniedError(errorMessage) {
  return errorMessage.includes(ShelbyErrorCodes.E_NOT_BLOB_OWNER) || errorMessage.includes(ShelbyErrorCodes.E_NOT_ADMIN) || errorMessage.includes("ENOT_AUTHORIZED") || errorMessage.includes("not authorized") || errorMessage.includes("permission denied");
}
function isBlobNotFoundError(errorMessage) {
  return errorMessage.includes(ShelbyErrorCodes.E_BLOB_NOT_FOUND);
}
function isBlobExpiredError(errorMessage) {
  return errorMessage.includes(ShelbyErrorCodes.E_BLOB_EXPIRED);
}

export {
  ShelbyErrorCodes,
  isBlobAlreadyExistsError,
  isAccessDeniedError,
  isBlobNotFoundError,
  isBlobExpiredError
};
