// @shelby-protocol/sdk/dist/chunk-NI6XBZLI.mjs
import {
  getShelbyIndexerClient
} from "./chunk-KYQRWJ3U.mjs";

// src/core/clients/ShelbyPlacementGroupClient.ts
var ShelbyPlacementGroupClient = class {
  indexer;
  /**
   * Creates a new ShelbyPlacementGroupClient.
   *
   * @param config - The client configuration object.
   *
   * @example
   * ```typescript
   * const pgClient = new ShelbyPlacementGroupClient({
   *   network: Network.SHELBYNET,
   * });
   * ```
   */
  constructor(config) {
    this.indexer = getShelbyIndexerClient(config);
  }
  /**
   * Retrieves placement group slots from the indexer.
   *
   * @param params.where (optional) - The where clause to filter slots by.
   * @param params.pagination (optional) - The pagination options.
   * @param params.orderBy (optional) - The order by clause to sort slots by.
   * @returns The placement group slots that match the filter.
   *
   * @example
   * ```typescript
   * // Get all active slots
   * const slots = await client.getPlacementGroupSlots({
   *   where: { status: { _eq: "active" } },
   * });
   * ```
   */
  async getPlacementGroupSlots(params) {
    const { limit, offset } = params.pagination ?? {};
    const { orderBy, where } = params;
    const { placement_group_slots } = await this.indexer.getPlacementGroupSlots(
      {
        where,
        limit,
        offset,
        orderBy
      }
    );
    return placement_group_slots.map(
      (slot) => ({
        placementGroup: slot.placement_group,
        slotIndex: Number(slot.slot_index),
        storageProvider: slot.storage_provider,
        status: slot.status,
        updatedAt: Number(slot.updated_at)
      })
    );
  }
  /**
   * Retrieves the total count of placement group slots from the indexer.
   *
   * @param params.where (optional) - The where clause to filter slots by.
   * @returns The count of placement group slots that match the filter.
   *
   * @example
   * ```typescript
   * // Get count of active slots
   * const count = await client.getPlacementGroupSlotsCount({
   *   where: { status: { _eq: "active" } },
   * });
   * ```
   */
  async getPlacementGroupSlotsCount(params) {
    const { where } = params;
    const { placement_group_slots_aggregate } = await this.indexer.getPlacementGroupSlotsCount({ where });
    return placement_group_slots_aggregate?.aggregate?.count ?? 0;
  }
};

export {
  ShelbyPlacementGroupClient
};
