// @shelby-protocol/sdk/dist/chunk-TET3DJEO.mjs
// src/core/operations/generated/sdk.ts
import gql from "graphql-tag";
var Blob_Activities_Select_Column = /* @__PURE__ */ ((Blob_Activities_Select_Column2) => {
  Blob_Activities_Select_Column2["BlobName"] = "blob_name";
  Blob_Activities_Select_Column2["EventIndex"] = "event_index";
  Blob_Activities_Select_Column2["EventType"] = "event_type";
  Blob_Activities_Select_Column2["Timestamp"] = "timestamp";
  Blob_Activities_Select_Column2["TransactionHash"] = "transaction_hash";
  Blob_Activities_Select_Column2["TransactionVersion"] = "transaction_version";
  return Blob_Activities_Select_Column2;
})(Blob_Activities_Select_Column || {});
var Blobs_Select_Column = /* @__PURE__ */ ((Blobs_Select_Column2) => {
  Blobs_Select_Column2["BlobCommitment"] = "blob_commitment";
  Blobs_Select_Column2["BlobName"] = "blob_name";
  Blobs_Select_Column2["CreatedAt"] = "created_at";
  Blobs_Select_Column2["ExpiresAt"] = "expires_at";
  Blobs_Select_Column2["IsDeleted"] = "is_deleted";
  Blobs_Select_Column2["IsWritten"] = "is_written";
  Blobs_Select_Column2["NumChunksets"] = "num_chunksets";
  Blobs_Select_Column2["Owner"] = "owner";
  Blobs_Select_Column2["PlacementGroup"] = "placement_group";
  Blobs_Select_Column2["Size"] = "size";
  Blobs_Select_Column2["SliceAddress"] = "slice_address";
  Blobs_Select_Column2["UpdatedAt"] = "updated_at";
  return Blobs_Select_Column2;
})(Blobs_Select_Column || {});
var Cursor_Ordering = /* @__PURE__ */ ((Cursor_Ordering2) => {
  Cursor_Ordering2["Asc"] = "ASC";
  Cursor_Ordering2["Desc"] = "DESC";
  return Cursor_Ordering2;
})(Cursor_Ordering || {});
var Order_By = /* @__PURE__ */ ((Order_By2) => {
  Order_By2["Asc"] = "asc";
  Order_By2["AscNullsFirst"] = "asc_nulls_first";
  Order_By2["AscNullsLast"] = "asc_nulls_last";
  Order_By2["Desc"] = "desc";
  Order_By2["DescNullsFirst"] = "desc_nulls_first";
  Order_By2["DescNullsLast"] = "desc_nulls_last";
  return Order_By2;
})(Order_By || {});
var Placement_Group_Slots_Select_Column = /* @__PURE__ */ ((Placement_Group_Slots_Select_Column2) => {
  Placement_Group_Slots_Select_Column2["PlacementGroup"] = "placement_group";
  Placement_Group_Slots_Select_Column2["SlotIndex"] = "slot_index";
  Placement_Group_Slots_Select_Column2["Status"] = "status";
  Placement_Group_Slots_Select_Column2["StorageProvider"] = "storage_provider";
  Placement_Group_Slots_Select_Column2["UpdatedAt"] = "updated_at";
  return Placement_Group_Slots_Select_Column2;
})(Placement_Group_Slots_Select_Column || {});
var Processor_Status_Select_Column = /* @__PURE__ */ ((Processor_Status_Select_Column2) => {
  Processor_Status_Select_Column2["LastSuccessVersion"] = "last_success_version";
  Processor_Status_Select_Column2["LastTransactionTimestamp"] = "last_transaction_timestamp";
  Processor_Status_Select_Column2["LastUpdated"] = "last_updated";
  return Processor_Status_Select_Column2;
})(Processor_Status_Select_Column || {});
var GetBlobsDocument = gql`
    query getBlobs($where: blobs_bool_exp, $orderBy: [blobs_order_by!], $limit: Int, $offset: Int) {
  blobs(where: $where, order_by: $orderBy, limit: $limit, offset: $offset) {
    owner
    blob_commitment
    blob_name
    created_at
    expires_at
    num_chunksets
    is_deleted
    is_written
    placement_group
    size
    updated_at
    slice_address
  }
}
    `;
var GetBlobActivitiesDocument = gql`
    query getBlobActivities($where: blob_activities_bool_exp, $orderBy: [blob_activities_order_by!], $limit: Int, $offset: Int) {
  blob_activities(
    where: $where
    order_by: $orderBy
    limit: $limit
    offset: $offset
  ) {
    blob_name
    event_index
    event_type
    transaction_hash
    transaction_version
    timestamp
  }
}
    `;
var GetBlobsCountDocument = gql`
    query getBlobsCount($where: blobs_bool_exp) {
  blobs_aggregate(where: $where) {
    aggregate {
      count
    }
  }
}
    `;
var GetBlobActivitiesCountDocument = gql`
    query getBlobActivitiesCount($where: blob_activities_bool_exp) {
  blob_activities_aggregate(where: $where) {
    aggregate {
      count
    }
  }
}
    `;
var GetTotalBlobsSizeDocument = gql`
    query getTotalBlobsSize($where: blobs_bool_exp) {
  blobs_aggregate(where: $where) {
    aggregate {
      sum {
        size
      }
    }
  }
}
    `;
var GetPlacementGroupSlotsDocument = gql`
    query getPlacementGroupSlots($where: placement_group_slots_bool_exp, $orderBy: [placement_group_slots_order_by!], $limit: Int, $offset: Int) {
  placement_group_slots(
    where: $where
    order_by: $orderBy
    limit: $limit
    offset: $offset
  ) {
    placement_group
    slot_index
    storage_provider
    status
    updated_at
  }
}
    `;
var GetPlacementGroupSlotsCountDocument = gql`
    query getPlacementGroupSlotsCount($where: placement_group_slots_bool_exp) {
  placement_group_slots_aggregate(where: $where) {
    aggregate {
      count
    }
  }
}
    `;
var GetProcessorStatusDocument = gql`
    query getProcessorStatus {
  processor_status {
    last_success_version
    last_transaction_timestamp
    last_updated
  }
}
    `;
var defaultWrapper = (action, _operationName, _operationType, _variables) => action();
function getSdk(client, withWrapper = defaultWrapper) {
  return {
    getBlobs(variables, requestHeaders, signal) {
      return withWrapper((wrappedRequestHeaders) => client.request({ document: GetBlobsDocument, variables, requestHeaders: { ...requestHeaders, ...wrappedRequestHeaders }, signal }), "getBlobs", "query", variables);
    },
    getBlobActivities(variables, requestHeaders, signal) {
      return withWrapper((wrappedRequestHeaders) => client.request({ document: GetBlobActivitiesDocument, variables, requestHeaders: { ...requestHeaders, ...wrappedRequestHeaders }, signal }), "getBlobActivities", "query", variables);
    },
    getBlobsCount(variables, requestHeaders, signal) {
      return withWrapper((wrappedRequestHeaders) => client.request({ document: GetBlobsCountDocument, variables, requestHeaders: { ...requestHeaders, ...wrappedRequestHeaders }, signal }), "getBlobsCount", "query", variables);
    },
    getBlobActivitiesCount(variables, requestHeaders, signal) {
      return withWrapper((wrappedRequestHeaders) => client.request({ document: GetBlobActivitiesCountDocument, variables, requestHeaders: { ...requestHeaders, ...wrappedRequestHeaders }, signal }), "getBlobActivitiesCount", "query", variables);
    },
    getTotalBlobsSize(variables, requestHeaders, signal) {
      return withWrapper((wrappedRequestHeaders) => client.request({ document: GetTotalBlobsSizeDocument, variables, requestHeaders: { ...requestHeaders, ...wrappedRequestHeaders }, signal }), "getTotalBlobsSize", "query", variables);
    },
    getPlacementGroupSlots(variables, requestHeaders, signal) {
      return withWrapper((wrappedRequestHeaders) => client.request({ document: GetPlacementGroupSlotsDocument, variables, requestHeaders: { ...requestHeaders, ...wrappedRequestHeaders }, signal }), "getPlacementGroupSlots", "query", variables);
    },
    getPlacementGroupSlotsCount(variables, requestHeaders, signal) {
      return withWrapper((wrappedRequestHeaders) => client.request({ document: GetPlacementGroupSlotsCountDocument, variables, requestHeaders: { ...requestHeaders, ...wrappedRequestHeaders }, signal }), "getPlacementGroupSlotsCount", "query", variables);
    },
    getProcessorStatus(variables, requestHeaders, signal) {
      return withWrapper((wrappedRequestHeaders) => client.request({ document: GetProcessorStatusDocument, variables, requestHeaders: { ...requestHeaders, ...wrappedRequestHeaders }, signal }), "getProcessorStatus", "query", variables);
    }
  };
}

export {
  Blob_Activities_Select_Column,
  Blobs_Select_Column,
  Cursor_Ordering,
  Order_By,
  Placement_Group_Slots_Select_Column,
  Processor_Status_Select_Column,
  GetBlobsDocument,
  GetBlobActivitiesDocument,
  GetBlobsCountDocument,
  GetBlobActivitiesCountDocument,
  GetTotalBlobsSizeDocument,
  GetPlacementGroupSlotsDocument,
  GetPlacementGroupSlotsCountDocument,
  GetProcessorStatusDocument,
  getSdk
};
