// @shelby-protocol/sdk/dist/chunk-XRLHLAOW.mjs
import {
  ClayErasureCodingProvider
} from "./chunk-OGKZ575S.mjs";
import {
  DEFAULT_CHUNK_SIZE_BYTES
} from "./chunk-JTXYKO3U.mjs";
import {
  ERASURE_CODE_PARAMS
} from "./chunk-AUQDI5BS.mjs";

// src/core/erasure/default.ts
var defaultProviderPromise;
function createDefaultErasureCodingProvider() {
  if (!defaultProviderPromise) {
    defaultProviderPromise = ClayErasureCodingProvider.create(
      defaultErasureCodingConfig()
    );
  }
  return defaultProviderPromise;
}
function defaultErasureCodingConfig() {
  const encoding = Number(process.env.SHELBY_ENCODING) || 0;
  if (encoding === 0) {
    return erasureCodingConfig16Total10Data13Helper();
  }
  return erasureCodingConfig4Total2Data3Helper();
}
function erasureCodingConfig16Total10Data13Helper() {
  return erasureCodingConfig(
    "ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */
  );
}
function erasureCodingConfig4Total2Data3Helper() {
  return erasureCodingConfig("ClayCode_4Total_2Data_3Helper" /* ClayCode_4Total_2Data_3Helper */);
}
function erasureCodingConfig(encodingScheme) {
  return {
    erasure_n: ERASURE_CODE_PARAMS[encodingScheme].erasure_n,
    erasure_k: ERASURE_CODE_PARAMS[encodingScheme].erasure_k,
    erasure_d: ERASURE_CODE_PARAMS[encodingScheme].erasure_d,
    chunkSizeBytes: DEFAULT_CHUNK_SIZE_BYTES,
    enumIndex: ERASURE_CODE_PARAMS[encodingScheme].enumIndex
  };
}

export {
  createDefaultErasureCodingProvider,
  defaultErasureCodingConfig,
  erasureCodingConfig16Total10Data13Helper,
  erasureCodingConfig4Total2Data3Helper
};
