// @shelby-protocol/sdk/dist/chunk-BUWAD67L.mjs
import {
  SenderBuiltMicropayment
} from "./chunk-TVLV7C74.mjs";
import {
  getShelbyIndexerClient
} from "./chunk-KYQRWJ3U.mjs";
import {
  getAptosConfig
} from "./chunk-IKRWBDG5.mjs";
import {
  normalizeAddress
} from "./chunk-UEZNZBJO.mjs";
import {
  MICROPAYMENTS_DEPLOYER
} from "./chunk-7OV5ZYW6.mjs";

// src/core/clients/ShelbyMicropaymentChannelClient.ts
import {
  AccountAddress,
  Aptos,
  MoveVector
} from "@aptos-labs/ts-sdk";
var ShelbyMicropaymentChannelClient = class _ShelbyMicropaymentChannelClient {
  aptos;
  deployer;
  indexer;
  /**
   * The ShelbyMicropaymentChannelClient is used to interact with the Micropayment contract on the Aptos blockchain. This
   * includes functions for the full lifecycle of micropayment channels: creation, withdrawals, expiration extensions, etc.
   *
   * @param config - The client configuration object.
   * @param config.network - The Shelby network to use.
   *
   * @example
   * ```typescript
   * const client = new ShelbyMicropaymentChannelClient({
   *   aptos: {
   *     network: Network.SHELBYNET,
   *     clientConfig: {
   *       API_KEY: "AG-***",
   *     },
   *   },
   * });
   * ```
   */
  constructor(config) {
    this.aptos = new Aptos(getAptosConfig(config));
    this.deployer = config.deployer ?? AccountAddress.fromString(MICROPAYMENTS_DEPLOYER);
    this.indexer = getShelbyIndexerClient(config);
  }
  /**
   * Calls initialize function so caller can thereafter create micropayment channels.
   *
   * @param params.sender - The account that will be sending funds to various receivers in micropayment channels.
   * @param params.options - Optional transaction building options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   *
   * const { transaction } = await client.initializePaymentChannels({
   *   sender: sender,
   * });
   * ```
   */
  async initializePaymentChannels(params) {
    const transaction = await this.aptos.transaction.build.simple({
      options: params.options,
      data: {
        function: `${(this.deployer ?? MICROPAYMENTS_DEPLOYER).toString()}::micropayments::initialize_payment_channels`,
        functionArguments: []
      },
      sender: params.sender.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.sender,
        transaction
      })
    };
  }
  /**
   * Creates a micropayment channel to start a channel bweteen a sender and a receiver for a specific asset.
   *
   * @param params.deployer - Optional deployer account address. Defaults to SHELBY_DEPLOYER.
   * @param params.sender - The account that will be sending funds.
   * @param params.receiver - The account that will be receiving funds.
   * @param params.expirationMicros - The expiration of the micropayment channel.
   * @param params.depositAmount - The amount of the asset locked to the channel.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   * @param params.options - Optional transaction building options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   *
   * const { transaction } = await client.createChannel({
   *   sender: sender,
   *   receiver: receiver,
   *   expirationMicros: Date.now() * 1000 + 86400_000_000, // 1 day from now in microseconds
   *   depositAmount: 0.001,
   *   fungibleAssetAddress: fungibleAssetAddress,
   * });
   * ```
   */
  async createChannel(params) {
    const transaction = await this.aptos.transaction.build.simple({
      options: params.options,
      data: _ShelbyMicropaymentChannelClient.makeCreateMicropaymentChannelPayload(
        {
          deployer: this.deployer,
          receiver: params.receiver,
          expirationMicros: params.expirationMicros,
          depositAmount: params.depositAmount,
          fungibleAssetAddress: params.fungibleAssetAddress,
          publicKey: params.publicKey ?? params.sender.publicKey.toUint8Array()
        }
      ),
      sender: params.sender.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.sender,
        transaction
      })
    };
  }
  /**
   * Creates a micropayment channel creation payload to start a channel between a sender and a receiver for a specific asset.
   * This is a static helper method for constructing the Move function call payload.
   *
   * @param params.deployer - Optional deployer account address. Defaults to MICROPAYMENTS_DEPLOYER.
   * @param params.sender - The account that will be sending funds.
   * @param params.receiver - The account that will be receiving funds.
   * @param params.expirationMicros - The expiration of the micropayment channel.
   * @param params.depositAmount - The amount of the asset locked to the channel.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   * @param params.publicKey - The Ed25519 public key bytes for the channel.
   *
   * @returns An Aptos transaction payload data object for the create_channel Move function.
   */
  static makeCreateMicropaymentChannelPayload(params) {
    return {
      function: `${(params.deployer ?? MICROPAYMENTS_DEPLOYER).toString()}::micropayments::create_channel`,
      functionArguments: [
        params.receiver,
        params.expirationMicros,
        params.depositAmount,
        params.fungibleAssetAddress,
        MoveVector.U8(params.publicKey)
      ]
    };
  }
  /**
   * Withdraws funds from a micropayment channel as the sender.
   *
   * @param params.sender - The account that is withdrawing funds.
   * @param params.receiver - The account the channel is configured to send funds to.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   * @param params.options - Optional transaction generation options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   * const { transaction } = await client.senderWithdraw({
   *   sender: sender,
   *   receiver: receiver,
   *   fungibleAssetAddress: fungibleAssetAddress,
   * });
   * ```
   */
  async senderWithdraw(params) {
    const transaction = await this.aptos.transaction.build.simple({
      options: params.options,
      data: _ShelbyMicropaymentChannelClient.createSenderWithdrawPayload({
        deployer: this.deployer,
        receiver: params.receiver,
        fungibleAssetAddress: params.fungibleAssetAddress
      }),
      sender: params.sender.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.sender,
        transaction
      })
    };
  }
  /**
   * Creates a static payload for the sender_withdraw Move function.
   * This is a helper method for constructing the transaction payload without signing.
   *
   * @param params.deployer - Optional deployer account address. Defaults to MICROPAYMENTS_DEPLOYER.
   * @param params.receiver - The account address of the receiver.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   *
   * @returns An Aptos transaction payload data object for the sender_withdraw Move function.
   */
  static createSenderWithdrawPayload(params) {
    return {
      function: `${(params.deployer ?? MICROPAYMENTS_DEPLOYER).toString()}::micropayments::sender_withdraw`,
      functionArguments: [params.receiver, params.fungibleAssetAddress]
    };
  }
  /**
   * Creates a micropayment that can be sent to a receiver.
   * The sender signs a WithdrawApproval message that authorizes the receiver
   * to withdraw funds from the micropayment channel.
   *
   * @param params.sender - The account that owns the payment channel and signs the approval.
   * @param params.receiver - The account address that will receive funds.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   * @param params.amount - The cumulative amount to authorize (not incremental).
   * @param params.paymentChannelId - The id of the payment channel.
   * @param params.sequenceNumber - The sequence number of the micropayment. Must be >= channel's next_withdrawn_sequence_number.
   *
   * @returns A SenderBuiltMicropayment containing the signed approval.
   *
   * @example
   * ```typescript
   * const senderBuiltMicropayment = client.createMicropayment({
   *   sender: senderAccount,
   *   receiver: receiverAddress,
   *   fungibleAssetAddress: fungibleAssetAddress,
   *   amount: 100,
   *   paymentChannelId: channelId,
   *   sequenceNumber: 1,
   * });
   * ```
   */
  createMicropayment(params) {
    const message = SenderBuiltMicropayment.buildSignedMessage({
      deployer: this.deployer,
      sender: params.sender.accountAddress,
      receiver: params.receiver,
      fungibleAssetAddress: params.fungibleAssetAddress,
      amount: params.amount,
      paymentChannelId: params.paymentChannelId,
      sequenceNumber: params.sequenceNumber
    });
    const signature = params.sender.sign(message);
    return new SenderBuiltMicropayment(
      params.sender.accountAddress,
      params.receiver,
      BigInt(params.paymentChannelId),
      BigInt(params.amount),
      params.fungibleAssetAddress,
      BigInt(params.sequenceNumber),
      params.sender.publicKey.toUint8Array(),
      signature.toUint8Array(),
      this.deployer
    );
  }
  /**
   * Creates a static payload for the receiver_withdraw Move function.
   * This is a helper method for constructing the transaction payload.
   *
   * @param params.deployer - Optional deployer account address. Defaults to MICROPAYMENTS_DEPLOYER.
   * @param params.sender - The sender's account address.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   * @param params.amount - The cumulative amount authorized to withdraw.
   * @param params.paymentChannelId - The payment channel id.
   * @param params.sequenceNumber - The sequence number for this withdrawal.
   * @param params.signature - The Ed25519 signature bytes.
   *
   * @returns An Aptos transaction payload data object for the receiver_withdraw Move function.
   */
  static createMicropaymentTransactionPayload(params) {
    return {
      function: `${(params.deployer ?? MICROPAYMENTS_DEPLOYER).toString()}::micropayments::receiver_withdraw`,
      functionArguments: [
        params.sender,
        params.fungibleAssetAddress,
        params.amount,
        params.paymentChannelId,
        params.sequenceNumber,
        MoveVector.U8(params.signature)
      ]
    };
  }
  /**
   * Withdraws funds from a micropayment channel as the receiver.
   * The receiver submits the sender's signed approval to claim funds.
   *
   * @param params.receiver - The receiver account that will sign and submit the transaction.
   * @param params.micropayment - The sender's signed micropayment approval.
   * @param params.options - Optional transaction generation options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   * const { transaction } = await client.receiverWithdraw({
   *   receiver: receiverAccount,
   *   micropayment: senderBuiltMicropayment,
   * });
   * ```
   */
  async receiverWithdraw(params) {
    const transaction = await this.aptos.transaction.build.simple({
      options: params.options,
      data: _ShelbyMicropaymentChannelClient.createMicropaymentTransactionPayload(
        {
          deployer: this.deployer,
          sender: params.micropayment.sender,
          fungibleAssetAddress: params.micropayment.fungibleAssetAddress,
          amount: params.micropayment.amount,
          paymentChannelId: params.micropayment.paymentChannelId,
          sequenceNumber: params.micropayment.sequenceNumber,
          signature: params.micropayment.signature
        }
      ),
      sender: params.receiver.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.receiver,
        transaction
      })
    };
  }
  /**
   * Extends the expiration time of a micropayment channel.
   *
   * @param params.sender - The owner of the account.
   * @param params.receiver - The account address of the receiver.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   * @param params.newExpirationMicros - The new expiration time in microseconds.
   * @param params.options - Optional transaction generation options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   * const { transaction } = await client.extendExpirationTime({
   *   sender: sender,
   *   receiver: receiverAddress,
   *   fungibleAssetAddress: fungibleAssetAddress,
   *   newExpirationMicros: Date.now() * 1000 + 172800_000_000, // 2 days from now
   * });
   * ```
   */
  async extendExpirationTime(params) {
    const transaction = await this.aptos.transaction.build.simple({
      options: params.options,
      data: _ShelbyMicropaymentChannelClient.makeExtendExpirationTimePayload({
        deployer: this.deployer,
        receiver: params.receiver,
        fungibleAssetAddress: params.fungibleAssetAddress,
        newExpirationMicros: params.newExpirationMicros
      }),
      sender: params.sender.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.sender,
        transaction
      })
    };
  }
  /**
   * Creates a payload to extend the expiration time on a micropayment channel.
   * This is a static helper method for constructing the Move function call payload.
   *
   * @param params.deployer - Optional deployer account address. Defaults to MICROPAYMENTS_DEPLOYER.
   * @param params.receiver - The account that will be receiving funds.
   * @param params.expirationMicros - The new expiration of the micropayment channel.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   *
   * @returns An Aptos transaction payload data object for the extend_expiration_time Move function.
   */
  static makeExtendExpirationTimePayload(params) {
    return {
      function: `${(params.deployer ?? MICROPAYMENTS_DEPLOYER).toString()}::micropayments::extend_expiration_time`,
      functionArguments: [
        params.receiver,
        params.fungibleAssetAddress,
        params.newExpirationMicros
      ]
    };
  }
  /**
   * Increases the deposit amount locked in a micropayment channel.
   *
   * @param params.sender - The account that is increasing the deposit.
   * @param params.receiver - The account address of the receiver.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   * @param params.additionalAmount - The additional amount to deposit.
   * @param params.options - Optional transaction generation options.
   *
   * @returns An object containing the pending transaction.
   *
   * @example
   * ```typescript
   * const { transaction } = await client.increaseDepositAmount({
   *   sender: sender,
   *   receiver: receiverAddress,
   *   fungibleAssetAddress: fungibleAssetAddress,
   *   additionalAmount: 500,
   * });
   * ```
   */
  async increaseDepositAmount(params) {
    const transaction = await this.aptos.transaction.build.simple({
      options: params.options,
      data: _ShelbyMicropaymentChannelClient.makeIncreaseDepositAmountPayload({
        deployer: this.deployer,
        receiver: params.receiver,
        fungibleAssetAddress: params.fungibleAssetAddress,
        additionalAmount: params.additionalAmount
      }),
      sender: params.sender.accountAddress
    });
    return {
      transaction: await this.aptos.signAndSubmitTransaction({
        signer: params.sender,
        transaction
      })
    };
  }
  /**
   * Creates a payload to increase the deposit amount on a micropayment channel.
   * This is a static helper method for constructing the Move function call payload.
   *
   * @param params.deployer - Optional deployer account address. Defaults to MICROPAYMENTS_DEPLOYER.
   * @param params.receiver - The account that will be receiving funds.
   * @param params.fungibleAssetAddress - The account address of the fungible asset.
   * @param params.additionalAmount - The amount of the fungible asset to add to the channel's balance.
   *
   * @returns An Aptos transaction payload data object for the extend_expiration_time Move function.
   */
  static makeIncreaseDepositAmountPayload(params) {
    return {
      function: `${(params.deployer ?? MICROPAYMENTS_DEPLOYER).toString()}::micropayments::increase_deposit_amount`,
      functionArguments: [
        params.receiver,
        params.fungibleAssetAddress,
        params.additionalAmount
      ]
    };
  }
  /**
   * Retrieves the vector of channels owned by the sender, optionally
   * filtered by the receiver.
   *
   * @param params.sender - The account namespace owning the micropayment channels.
   * @param params.receiver? - An optional receiver account address.
   * @returns A vector of ChannelInfo objects.
   *
   * @example
   * ```typescript
   * const channelInfoVec = await client.getChannelInfo({
   *   sender: AccountAddress.fromString("0x1"),
   *   receiver: AccountAddress.fromString("0x2"),
   * });
   * ```
   */
  async getChannelInfo(params) {
    const functionName = params.receiver ? "get_all_channels_from_sender_to_receiver" : "get_all_channels_from_sender";
    const functionArguments = [params.sender];
    if (params.receiver) {
      functionArguments.push(params.receiver);
    }
    const rawArray = await this.aptos.view({
      payload: {
        function: `${this.deployer.toString()}::micropayments::${functionName}`,
        functionArguments
      }
    });
    const channelInfoArray = rawArray[0];
    return channelInfoArray.map(
      (channelInfo) => {
        const publicKeyHex = channelInfo.public_key.startsWith("0x") ? channelInfo.public_key.slice(2) : channelInfo.public_key;
        const publicKey = new Uint8Array(
          publicKeyHex.match(/.{1,2}/g)?.map((byte) => Number.parseInt(byte, 16)) ?? []
        );
        return {
          sender: params.sender,
          receiver: params.receiver,
          fungibleAssetAddress: normalizeAddress(
            channelInfo.asset_metadata.inner
          ),
          balance: Number(channelInfo.balance),
          creationMicros: Number(channelInfo.creation_micros),
          expirationMicros: Number(channelInfo.expiration_micros),
          nextWithdrawnSequenceNumber: Number(
            channelInfo.next_withdrawn_sequence_number
          ),
          paymentChannelId: Number(channelInfo.payment_channel_id),
          receiverWithdrawnAmount: Number(
            channelInfo.receiver_withdrawn_amount
          ),
          publicKey
        };
      }
    );
  }
};

export {
  ShelbyMicropaymentChannelClient
};
