// @shelby-protocol/sdk/dist/chunk-7OV5ZYW6.mjs
// src/core/constants.ts
import { Network } from "@aptos-labs/ts-sdk";
var DEFAULT_PROJECT_NAME = "shelby";
var DEFAULT_PROJECT_DESCRIPTION = "High performance, decentralized storage";
var NetworkToShelbyRPCBaseUrl = {
  [Network.SHELBYNET]: "https://api.shelbynet.shelby.xyz/shelby",
  [Network.NETNA]: void 0,
  [Network.DEVNET]: void 0,
  [Network.TESTNET]: "https://api.testnet.shelby.xyz/shelby",
  [Network.MAINNET]: void 0,
  [Network.LOCAL]: void 0,
  [Network.CUSTOM]: void 0
};
var NetworkToShelbyBlobIndexerBaseUrl = {
  [Network.SHELBYNET]: "https://api.shelbynet.aptoslabs.com/nocode/v1/public/cmforrguw0042s601fn71f9l2/v1/graphql",
  [Network.NETNA]: void 0,
  [Network.DEVNET]: void 0,
  [Network.TESTNET]: "https://api.testnet.aptoslabs.com/nocode/v1/public/cmlfqs5wt00qrs601zt5s4kfj/v1/graphql",
  [Network.MAINNET]: void 0,
  [Network.LOCAL]: void 0,
  [Network.CUSTOM]: void 0
};
var NetworkToGasStationBaseUrl = {
  [Network.SHELBYNET]: "https://api.shelbynet.shelby.xyz/gs/v1",
  [Network.NETNA]: void 0,
  [Network.DEVNET]: void 0,
  [Network.TESTNET]: "https://api.testnet.shelby.xyz/gs/v1",
  [Network.MAINNET]: void 0,
  [Network.LOCAL]: void 0,
  [Network.CUSTOM]: void 0
};
var SHELBY_DEPLOYER = "0x85fdb9a176ab8ef1d9d9c1b60d60b3924f0800ac1de1cc2085fb0b8bb4988e6a";
var MICROPAYMENTS_DEPLOYER = "0x1ae7275148bf6ef742b658fd9cbcc2e094201606f4a7bc707bab0201da8043ee";
var TOKEN_DEPLOYER = "0x33009e852be7f93762dd0bf303383c2cb2c5cab7a30d8238ca5f9f177ae75124";
var TOKEN_OBJECT_ADDRESS = "0x249f5c642a63885ff88a5113b3ba0079840af5a1357706f8c7f3bfc5dd12511f";
var SHELBYUSD_TOKEN_ADDRESS = TOKEN_OBJECT_ADDRESS;
var SHELBYUSD_TOKEN_MODULE = "shelby_usd";
var SHELBYUSD_TOKEN_NAME = "ShelbyUSD";
var SHELBYUSD_FA_METADATA_ADDRESS = "0x1b18363a9f1fe5e6ebf247daba5cc1c18052bb232efdc4c50f556053922d98e1";

export {
  DEFAULT_PROJECT_NAME,
  DEFAULT_PROJECT_DESCRIPTION,
  NetworkToShelbyRPCBaseUrl,
  NetworkToShelbyBlobIndexerBaseUrl,
  NetworkToGasStationBaseUrl,
  SHELBY_DEPLOYER,
  MICROPAYMENTS_DEPLOYER,
  TOKEN_DEPLOYER,
  TOKEN_OBJECT_ADDRESS,
  SHELBYUSD_TOKEN_ADDRESS,
  SHELBYUSD_TOKEN_MODULE,
  SHELBYUSD_TOKEN_NAME,
  SHELBYUSD_FA_METADATA_ADDRESS
};
