// @shelby-protocol/sdk/dist/chunk-Z4FZ7W6L.mjs
// src/core/layout.ts
import { z } from "zod";
var BlobNameSchema = z.string().min(1, "Blob name path parameter cannot be empty.").max(1024, "Blob name cannot exceed 1024 characters.").refine((name) => !name.endsWith("/"), {
  message: "Blob name cannot end with a slash"
});
var ChunkKey = class {
  constructor(account, blobName, chunksetIdx, chunkIdx, config) {
    this.account = account;
    this.blobName = blobName;
    this.chunksetIdx = chunksetIdx;
    this.chunkIdx = chunkIdx;
    this.config = config;
    if (chunkIdx >= config.erasure_n) {
      throw new Error(
        `Cannot create a chunk with idx ${chunkIdx}. M+K=${config.erasure_n}`
      );
    }
  }
  key() {
    return `${this.account.toString()}.${this.blobName}.${this.chunksetIdx}.${this.chunkIdx}`;
  }
  // Returns the range in the blob that this key represents, or "parity" if this is parity chunk.
  range() {
    if (this.chunkIdx >= this.config.erasure_k) {
      return "parity";
    }
    const chunksetStart = this.chunksetIdx * this.config.chunkSizeBytes * this.config.erasure_k;
    const chunkStart = chunksetStart + this.chunkIdx * this.config.chunkSizeBytes;
    return {
      start: chunkStart,
      end: chunkStart + this.config.chunkSizeBytes
    };
  }
};

export {
  BlobNameSchema,
  ChunkKey
};
