// @shelby-protocol/sdk/dist/chunk-SSE4MFNN.mjs
import {
  getAptosConfig
} from "./chunk-IKRWBDG5.mjs";
import {
  normalizeAddress
} from "./chunk-UEZNZBJO.mjs";
import {
  SHELBY_DEPLOYER
} from "./chunk-7OV5ZYW6.mjs";

// src/core/clients/ShelbyMetadataClient.ts
import {
  AccountAddress,
  Aptos,
  Hex
} from "@aptos-labs/ts-sdk";
function parseStorageProviderState(raw) {
  switch (raw.__variant__) {
    case "Active":
      return {
        variant: "Active",
        quota: raw.quota.value,
        stakeAtStartOfStakingEpoch: raw.stake_at_start_of_staking_epoch,
        faulty: raw.faulty,
        leaving: raw.leaving
      };
    case "Waitlisted":
      return {
        variant: "Waitlisted"
      };
    case "Frozen":
      return {
        variant: "Frozen",
        frozenReason: raw.frozen_reason,
        frozenFrom: raw.frozen_from,
        frozenTill: raw.frozen_till
      };
  }
}
var ShelbyMetadataClient = class {
  aptos;
  deployer;
  /**
   * The ShelbyMetadataClient is used to interact with the Shelby contract on the Aptos blockchain. This
   * includes functions like gathering basic details about the Shelby system, including storage provider information.
   *
   * @param config.aptos.config - The Aptos config.
   * @param config.shelbyDeployer - The deployer account address of the Shelby contract. If not provided, the default deployer address will be used.
   *
   * @example
   * ```typescript
   * const aptos = new Aptos(new AptosConfig({ network: Network.TESTNET }));
   * const metadataClient = new ShelbyMetadataClient({ aptos });
   * ```
   */
  constructor(config) {
    this.aptos = new Aptos(getAptosConfig(config));
    this.deployer = config.deployer ?? AccountAddress.fromString(SHELBY_DEPLOYER);
  }
  /**
   * Retrieves storage provider list from the blockchain.
   *
   * @returns A list of storage providers, or empty array if none exist.
   *
   * @example
   * ```typescript
   * const spList = await client.getStorageProviders();
   * ```
   */
  async getStorageProviders() {
    try {
      const rawMetadata = await this.aptos.view({
        payload: {
          function: `${this.deployer.toString()}::storage_provider_registry::get_all_storage_providers`,
          functionArguments: []
        }
      });
      const metadata = rawMetadata[0];
      return metadata.map((provider) => ({
        address: normalizeAddress(provider.address),
        ipAddress: provider.ip_address,
        port: provider.port,
        blsPublicKey: Hex.fromHexInput(provider.bls_public_key).toUint8Array(),
        availabilityZone: provider.availability_zone,
        state: parseStorageProviderState(provider.state)
      }));
    } catch (error) {
      if (error instanceof Error && // Depending on the network, the error message may show up differently.
      error.message?.includes("sub_status: Some(404)")) {
        return [];
      }
      throw error;
    }
  }
  /**
   * Retrieves the list of placement group addresses.
   *
   * @returns The placement group address list, or an empty array if none exist.
   *
   * @example
   * ```typescript
   * const pgList = await client.getPlacementGroupAddresses();
   * ```
   */
  async getPlacementGroupAddresses() {
    try {
      const pgSizeMetadata = await this.aptos.view({
        payload: {
          function: `${this.deployer.toString()}::placement_group_registry::get_number_of_placement_groups`,
          functionArguments: []
        }
      });
      const finalPlacementGroupIndex = pgSizeMetadata[0] - 1;
      const addressMetadataArray = await this.aptos.view({
        payload: {
          function: `${this.deployer.toString()}::placement_group_registry::get_placement_group_addresses`,
          functionArguments: [0, finalPlacementGroupIndex]
        }
      });
      const metadata = addressMetadataArray[0];
      return metadata.map((pg) => normalizeAddress(pg));
    } catch (error) {
      if (error instanceof Error && // Depending on the network, the error message may show up differently.
      (error.message?.includes("sub_status: Some(404)") || error.message?.includes("E_PLACEMENT_GROUP_NOT_FOUND"))) {
        return [];
      }
      throw error;
    }
  }
  /**
   * Retrieves the list of slice addresses.
   *
   * @returns The slice group list, or an empty array if none exist.
   *
   * @example
   * ```typescript
   * const pgList = await client.getSliceAddresses();
   * ```
   */
  async getSliceAddresses() {
    try {
      const sliceSizeMetadata = await this.aptos.view({
        payload: {
          function: `${this.deployer.toString()}::slice_registry::get_number_of_slices`,
          functionArguments: []
        }
      });
      const finalSliceIndex = sliceSizeMetadata[0] - 1;
      const addressMetadataArray = await this.aptos.view({
        payload: {
          function: `${this.deployer.toString()}::slice_registry::get_slice_addresses`,
          functionArguments: [0, finalSliceIndex]
        }
      });
      const metadata = addressMetadataArray[0];
      return metadata.map((slice) => normalizeAddress(slice));
    } catch (error) {
      if (error instanceof Error && // Depending on the network, the error message may show up differently.
      (error.message?.includes("sub_status: Some(404)") || error.message?.includes("E_SLICE_NOT_FOUND"))) {
        return [];
      }
      throw error;
    }
  }
  /**
   * Gets the placement group address for a slice.
   *
   * @param sliceAddress - The address of the slice account.
   * @returns The placement group address as a string.
   */
  async getPlacementGroupAddressForSlice(sliceAddress) {
    const sliceMetadata = await this.aptos.view({
      payload: {
        function: `${this.deployer.toString()}::slice::get_slice_info`,
        functionArguments: [sliceAddress.toString()]
      }
    });
    return sliceMetadata[0].placement_group_assignments[0].placement_group_address;
  }
  /**
   * Retrieves the designated storage providers for a slice.
   *
   * Designated SPs are those appointed to store data for their slots:
   * - Active: Currently serving data
   * - Receiving: Receiving data during slot transfer
   * - Repairing: Repairing data after crash or failed transfer
   * - Reconstructing: Reconstructing data if the repair fails
   *
   * @param params.account - The address of the slice account.
   * @returns An array where result[i] is the designated SP for slot i, or null if no SP is designated.
   *
   * @example
   * ```typescript
   * const providers = await client.getDesignatedStorageProvidersForSlice({ account: sliceAddress });
   * ```
   */
  async getDesignatedStorageProvidersForSlice(params) {
    const placementGroupAddress = await this.getPlacementGroupAddressForSlice(
      params.account
    );
    const rawMetadata = await this.aptos.view({
      payload: {
        function: `${this.deployer.toString()}::placement_group::get_designated_storage_providers`,
        functionArguments: [placementGroupAddress]
      }
    });
    const providers = rawMetadata[0];
    return providers.map(
      (opt) => opt.vec.length > 0 ? normalizeAddress(opt.vec[0]) : null
    );
  }
  /**
   * Retrieves the active storage providers for a slice.
   *
   * Active SPs have a complete copy of the data for the slot, and are not in any data transfer/repair/reconstruction phase.
   * Active SP can be audited for the data it contains.
   * Each slot has at most one active SP.
   *
   * @param params.account - The address of the slice account.
   * @returns An array where result[i] is the active SP for slot i, or null if no SP is active.
   *
   * @example
   * ```typescript
   * const providers = await client.getActiveStorageProvidersForSlice({ account: sliceAddress });
   * ```
   */
  async getActiveStorageProvidersForSlice(params) {
    const placementGroupAddress = await this.getPlacementGroupAddressForSlice(
      params.account
    );
    const rawMetadata = await this.aptos.view({
      payload: {
        function: `${this.deployer.toString()}::placement_group::get_active_storage_providers`,
        functionArguments: [placementGroupAddress]
      }
    });
    const providers = rawMetadata[0];
    return providers.map(
      (opt) => opt.vec.length > 0 ? normalizeAddress(opt.vec[0]) : null
    );
  }
  /**
   * Retrieves the serving storage providers for a slice.
   *
   * Serving SPs are those that can respond to read requests. The serving logic is:
   * - If an Active SP exists for a slot: Only the Active SP is serving
   * - If no Active SP (transition in progress): Both Designated and Vacating SPs serve
   *
   * Each slot may have multiple serving SPs during transitions.
   *
   * @param params.account - The address of the slice account.
   * @returns An array where result[i] contains the serving SPs for slot i.
   *
   * @example
   * ```typescript
   * const providers = await client.getServingStorageProvidersForSlice({ account: sliceAddress });
   * ```
   */
  async getServingStorageProvidersForSlice(params) {
    const placementGroupAddress = await this.getPlacementGroupAddressForSlice(
      params.account
    );
    const rawMetadata = await this.aptos.view({
      payload: {
        function: `${this.deployer.toString()}::placement_group::get_serving_storage_providers`,
        functionArguments: [placementGroupAddress]
      }
    });
    const providers = rawMetadata[0];
    return providers.map(
      (slotProviders) => slotProviders.map((addr) => normalizeAddress(addr))
    );
  }
};

export {
  ShelbyMetadataClient
};
